### PR TITLE
lightnovel

### LINKED ISSUE #873: Add Ippotranslations
### Source name

Ippotranslations

### Source link

https://ippotranslations.com/

### Language

English

### Other details

_No response_

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a title with source name.
- [X] I have checked that the extension does not already exist by searching the [GitHub repository](https://github.com/LNReader/lnreader/) and verified it does not appear in the code base.
- [X] I will fill out all of the requested information in this form.

### LINKED ISSUE #935: New domain name
### Source name

CPUnovel

### Source link

https://cpunovel.com/

### Language

English

### Other details

Well, it's the new domain name of Daotranslate 

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a title with source name.
- [X] I have checked that the extension does not already exist by searching the [GitHub repository](https://github.com/LNReader/lnreader/) and verified it does not appear in the code base.
- [X] I will fill out all of the requested information in this form.

### LINKED ISSUE #960: Sub-Titles are incorrectly displayed
### Source information

Ranobelib

### Steps to reproduce

Screenshot_2024-05-28-16-33-40-91_3c1c179116822eec69c21805387f0489
screen from random chapter from classroom of elite

### Expected behavior

for example in rulate:
Screenshot_2024-05-28-16-39-38-24_3c1c179116822eec69c21805387f0489

### Actual behavior

incorrectly displayed

### LNReader version

2.0.0(i have used last file from build actions)

### Android version

Android 13

### Other details

_No response_

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] I have updated the app to version **[1.1.19](https://github.com/LNReader/lnreader/releases/latest)**.
- [X] I have updated all installed extensions.
- [X] If this is an issue with the app itself, I should be opening an issue in the [app repository](https://github.com/LNReader/lnreader/issues/new/choose).
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by Rider21**
Added cpunovel plugin (closes #935)
Added Ippotranslations plugin (closes #873)
Plugin ellotl moved to lightnovelwp
Fixed Ranobelib plugin (closes #960)

### CODE REVIEW

**Comment by Batorian** [comment]
I wrote a separate Plugin for ElloTL in order to make it work with `htmlparser2`. Multisrc would use cheerio.

**Comment by Rider21** [comment]
I think in `react-native` they will be the same in terms of execution time.
But I will rewrite lightnovelwp to htmlparser2 later.

File: scripts/multisrc/lightnovelwp/sources.json [hunk 1]
```diff
@@ -176,5 +176,32 @@
...
+    }
+  },
+  {
+    "id": "cpunovel",
+    "sourceSite": "https://cpunovel.com/",
+    "sourceName": "CPUnovel",
+    "options": {
+      "lang": "English",
+      "reverseChapters": true
+    }
```
Comment (error7404) [inline scripts/multisrc/lightnovelwp/sources.json:205]: Cpunovel isn't a new site it's juste the new name of daotranslate

**Comment by nyagami** [comment]
> I think in `react-native` they will be the same in terms of execution time. But I will rewrite lightnovelwp to htmlparser2 later.

Nah, Cheerio keep whole DOM tree in memory which helps you to querySelector, and building that tree (load) consumes much resource which lead to lag in update library and global search

File: scripts/multisrc/lightnovelwp/sources.json (hunk 1 above)
Comment (Rider21) [reply scripts/multisrc/lightnovelwp/sources.json:205]: they have different pages of `popular novels`, `recent releases`,
I think it's a new site.

**Comment by Rider21** [comment]
Hmm, then it's probably better to use `native` regex, match, substr, etc.
It will be much faster as you will only search for what you need.

**Comment by nyagami** [comment]
Not really, 
With htmlparser2, Time complexity is O(n*k) where k is the number of switch case/if else statements, therefore you can consider as O(n).
Meanwhile with regex, it takes O(n*q) where q is the number of queries. Even worse if you use some bad regex

**Review by nyagami — APPROVED** [review]

**Pull request merged**